flink-optimizer/src/main/java/org/apache/flink/optimizer/plantranslate/JobGraphGenerator.java

```diff
@@ -480,6 +482,32 @@ public void postVisit(PlanNode node) {
 			if (node instanceof SourcePlanNode || node instanceof NAryUnionPlanNode || node instanceof SolutionSetPlanNode) {
 				return;
 			}
+
+			// if this is a blocking shuffle vertex, we add one IntermediateDataSetID to its predecessor and return
+			if (node instanceof SinkPlanNode) {
+				Object userCodeObject = node.getProgramOperator().getUserCodeWrapper().getUserCodeObject();
+				if (userCodeObject instanceof BlockingShuffleOutputFormat) {
+					PlanNode precedentNode = node.getInputs().iterator().next().getSource();
```

Comment:
Do we need to check `node.getInputs().iterator().hasNext()` before `getSource()`?

Reply:
Since it's a SinkPlanNode, it contains one and only one input, I can add some sanity check here.